### PR TITLE
Generate index without downloading the whole file from blob storage

### PR DESCRIPTION
**Pull request by wwwjn**
### Related issues
#4201 

### Checklist

* [ ] I've added a screenshot of the changes, if this is a frontend change
* [ ] I've added and/or updated tests, if this is a backend change
* [ ] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [ ] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

**Comment by epicfaace** [comment]
CI error: 
image

**Comment by wwwjn** [comment]
> CI error: image

Failed test: https://github.com/codalab/codalab-worksheets/actions/runs/4063782840/jobs/6996560672
All the filed test cases are because we can not know the actual file size for a GzipStream, so SQLiteTar can not set correct file size field.

**Comment by wwwjn** [comment]
Issue: The modified index gzip can not get the actual file size of a single file. But directories are not influenced.

Since we can not get the **original file size** of a single file, several important functions are influenced:
1. `get_file_size()`. For now, we just return the compressed size as the file size.
2. `summarize_file()`: Since the compressed file is not seek-able, we need to read() the whole file to should the summarize of file content. 
3. `OpenFile()` in file_util.py.  The `TarFileStream()` class in `OpenFile()` function call, **needs the original file size**.

A potential solution: 
-  If the user upload a single file, we put the file into a temporary folder, and upload as a folder (upload file / folder as a `contents.tar.gz`).  
    - but a defect of this method is: we need to modify functions that decides the path of the file, whether the file is dir / folder
    - Another defect is that, we already have a lot of bundle named as `contents.gz`, we need to be compatible with them

**Comment by wwwjn** [comment]
pack: Pack the file and compress
force_compression: speeds up by compress the file

File: codalab/worker/download_util.py [hunk 1]
```diff
@@ -260,7 +261,7 @@ def _get_info(path: str, depth: Union[int, float]) -> TargetInfo:
             # The entry returned by ratarmount for a single .gz file is not technically part of a tar archive
             # and has a name hardcoded as "contents," so we modify the type, name, and permissions of
             # the output accordingly.
-            return cast(
+            result = cast(
```
Comment (epicfaace) [inline codalab/worker/download_util.py]: undo this change

File: codalab/server/bundle_manager.py [hunk 2]
```diff
@@ -278,8 +280,11 @@ def _make_bundle(self, bundle):
                             un_tar_directory(fileobj, dependency_path, 'gz')
                         else:
                             fileobj = self._download_manager.stream_file(target, gzipped=False)
+
                             with open(dependency_path, 'wb') as f:
                                 shutil.copyfileobj(fileobj, f)
+                                # f.seek(0)
```
Comment (epicfaace) [inline codalab/server/bundle_manager.py]: rm these changes

File: tst.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+from codalab.worker.bundle_state import State
+from codalab.lib import file_util, path_util, zip_util
+from codalab.objects.bundle import Bundle
+from codalab.lib.zip_util import ARCHIVE_EXTS_DIR
+from codalab.lib.print_util import FileTransferProgress
+from codalab.worker.un_gzip_stream import BytesBuffer
+
+import indexed_gzip
+from codalab.lib.beam.SQLiteIndexedTar import SQLiteIndexedTar
+from ratarmountcore import FileInfo
```
Comment (epicfaace) [inline tst.py]: rm this file

File: tests/cli/test_cli.py [hunk 4]
```diff
@@ -2212,7 +2212,7 @@ def test_read(ctx):
 
         # Cat has everything.
         cat_output = _run_command([cl, 'cat', uuid + '/stdout'])
-        check_contains('5\n6\n7', cat_output)
+        check_contains('5\n6\n7', cat_output)  # HERE failed
```
Comment (epicfaace) [inline tests/cli/test_cli.py]: rm all extra comments

File: codalab/lib/upload_manager.py [hunk 5]
```diff
@@ -296,11 +327,41 @@ def write_fileobj(
...
+                            else output_fileobj.tell()
+                        )
+                        if self.is_client:
+                            self._client.update(
+                                'bundles/%s/contents/filesize/' % bundle_uuid, {'filesize': file_size},
+                            )
+                        else:  # directly update on server side
+                            update_file_size(bundle_path, file_size)
+                    except Exception as e:
+                        print("Do nothing here")
```
Comment (epicfaace) [inline codalab/lib/upload_manager.py]: raise the Exception

File: codalab/lib/__pycache__/upload_manager.cpython-37.pyc.140238849648560
Comment (epicfaace) [inline codalab/lib/__pycache__/upload_manager.cpython-37.pyc.140238849648560:1]: rm files in `__pycache__`

File: tests/cli/test_cli.py [hunk 6]
```diff
@@ -871,7 +871,7 @@ def test_upload1(ctx):
             _run_command([cl, 'work', worksheet_uuid])
             # expect to fail when we upload something more than 2k bytes
             check_contains(
-                'User disk quota exceeded',
+                'disk quota exceeded',
```
Comment (epicfaace) [inline tests/cli/test_cli.py]: Comment for why this change was made (from my conversation with @wwwjn ):


found the issue





[7:13](https://codalab.slack.com/archives/D03A56V872R/p1679440386207059)
Basically, Andrew’s latest change makes it so that it only checks and runs “user disk quota exceeded” every ITERATIONS_PER_DISK_CHECK * CHUNK_SIZE bytes (see upload_manage.py).
[7:13](https://codalab.slack.com/archives/D03A56V872R/p1679440396233269)
This means, every 32 MB
[7:13](https://codalab.slack.com/archives/D03A56V872R/p1679440406940839)
So any file you upload that is under 32 MB, it won’t trigger that code path and thus will cause the test to fail.
[7:13](https://codalab.slack.com/archives/D03A56V872R/p1679440417893019)
Probably the files are too small when using azure blob since they’re zipped
[7:14](https://codalab.slack.com/archives/D03A56V872R/p1679440446989859)
So I just changed the test like this and I think it should fix it: https://github.com/codalab/codalab-worksheets/pull/4370/commits/7741c941dd8098d189332edf82c2388785c0142b

File: codalab/worker/file_util.py [hunk 7]
```diff
@@ -312,19 +315,48 @@ def __init__(self, fileobj: IO[bytes]):
...
+            self.__size += len(data)
+            return data
+        except Exception as e:
+            logging.info("Error in GzipStream read() ", repr(e))
+            return None
 
     def close(self):
         self.__input.close()
 
+    def peek(self, num_bytes):
```
Comment (epicfaace) [inline codalab/worker/file_util.py]: Do we still need the changes in this file?

File: codalab/rest/bundles.py [hunk 8]
```diff
@@ -772,6 +774,41 @@ def _fetch_bundle_contents_info(uuid, path=''):
     return {'data': info}
 
 
+@patch(
+    '/bundles/<uuid:re:%s>/contents/filesize/' % spec_util.UUID_STR, name='update_bundle_file_size'
+)
+def _update_bundle_file_size(uuid):
+    """
+    This function is used to fix the file size field in the index.sqlite file.
```
Comment (epicfaace) [inline codalab/rest/bundles.py]: Can we update the comment here to explain when it's used?

File: tests/unit/server/upload_download_test.py [hunk 9]
```diff
@@ -71,6 +71,7 @@ def test_not_found(self):
 
     def check_file_target_contents(self, target):
         """Checks to make sure that the specified file has the contents 'hello world'."""
+        # This can not be checked, Since
```
Comment (epicfaace) [inline tests/unit/server/upload_download_test.py]: rm

**Review by epicfaace — APPROVED** [review]

**Comment by wwwjn** [comment]
This PR was closed mistakenly, so I create a new PR #4444 .
